### ISSUE TITLE
Two Group By With avg() and varPop()

### ISSUE THREAD

**Issue by ongbo** [ClickHouse/ClickHouse#38080]
This Local Table is real time write, there will be duplicate data.
```
CREATE TABLE test.one(
    date string,
    hour String,
    app String,
    dim1 String,
    dim2 String,
    user String,
    value UInt64
)
Engine = MergeTree
```

This materialized view is used to merge into user granularity.
```
CREATE MATERIALIZED VIEW test.one_materialized_view_v1 on cluster tt
(
    date string,
    hour String,
    app String,
    dim1 String,
    dim2 String,
    user String,
    value AggregateFunction(sum, UInt64)
)
ENGINE = AggregatingMergeTree
PARTITION BY ..
ORDER BY (..)
SETTINGS index_granularity = 8192 AS
SELECT
    date, hour,app, dime1, dim2, user
    sumState(value) AS value
FROM test.one
GROUP BY date, hour,app, dime1, dim2, user;
```

My query operation：
```
select
    date, hour,app, dime1, dim2,
    sum(valuex),
    avg(valuex),
    varPop(valuex)
from
(
select
    date, hour,app, dime1, dim2, user, sumMerge(value) AS valuex
from test.one_materialized_view_v1
where ..
group by date, hour,app, dime1, dim2, user
) t
group by date, hour,app, dime1, dim2;
```

This query is very slow, maybe 7 or 8 seconds. Subquery base on materialized view, so it query fastly.
I heard the Aggregation over aggregation is single threaded, So I rebuild materialized view like this：

```

CREATE MATERIALIZED VIEW test.one_materialized_view_v2 on cluster cluster_music_bi_single
(
   date string,
    hour String,
    app String,
    dim1 String,
    dim2 String,
    `value` AggregateFunction(sum,  Float64) ,
    `mean_value` AggregateFunction(avg, Float64),
    `var_value` AggregateFunction(varPop, Float64)
)
ENGINE = AggregatingMergeTree
PARTITION BY ...
ORDER BY (...)
SETTINGS index_granularity = 8192 AS(
SELECT
    date, hour,app, dime1, dim2,
    sumState(t.valuex) AS value,
    avgState(t.valuex) AS mean_value,
    varPopState(t.valuex) AS var_value
FROM 
(
    SELECT
        date, hour,app, dime1, dim2, userid,
        sum(index_value) AS valuex
    FROM test.one
    GROUP BY date, hour,app, dime1, dim2, userid
) AS t
GROUP BY date, hour,app, dime1, dim2
)
```

My new query as:
```
select
    date, hour,app, dime1, dim2,
    sumMerge(value) AS sum_value,
    avgMerge(mean_value) as mean_valuex,
    varPopMerge(var_value) as var_valuex
from test.one_materialized_view_v2
group by date, hour,app, dime1, dim2
```

it query fastly, And sum_value is correct, but mean_valuex and var_valuex is wrong. it will be more smaller than one_materialized_view_v1.
his requirement is to calculate the mean and variance of user granularity, so it cannot be directly aggregated, otherwise it will increase the denominator;
From this result, it can be seen that the SQL semantics of materialized views do not aggregate completely to user granularity, 
but rather that each part increases the denominator when varPopMerge is evaluated

**Comment by den-crane** [comment]
>but mean_valuex and var_valuex is wrong. 

But this is expected. 
It's a math problem, `avg(sum())` is not associative func.

Imagine you have user1 with 1 and 1, user2 with 2 and 2.
With 2 group by you calculate  `(1+1 + 2+2)/2)==3` , with single groupby `((1+1+2+2)/4)==1.5`

You have to use the first way "Two Group By".

**Comment by helifu** [comment]
Hi @den-crane ,

**Comment by den-crane** [comment]
https://en.wikipedia.org/wiki/Associative_property

**Comment by ongbo** [comment]
Hi @den-crane ,
Thank you very much for your reply,
The SQL semantics of this materialized view is the essence of the semantic essence of the user to aggregate to the user's granularity, 
and then calculate the average value and square of the user's particle size. At least it looks like this on the user. 
If the underlying implementation is AVG (SUM ())In terms of, the understanding of the user is relatively high, 
or when we can understand it, we can provide some solutions. 
The essence is the base nominal N of the AVG function. It can be controlled by ourselves.
The value of the state maintenanceer Map <user, value> is also OK.
In other words, can you provide another solution? After all, the Two Group By query method is a single thread in the outside query, which is too slow.
Thanks in advance. :)

**Comment by den-crane** [comment]
SELECT query in materialized view  HAS ZERO RELATION and ZERO IMPACT on AggregatingMergeTree.

It does not matter that you invented some megaquery in materialized view. 
Materialized view is an insert trigger. I suggest to validate your MatView by a mental experiment -- insert only one row at a time.
https://den-crane.github.io/Everything_you_should_know_about_materialized_views_commented.pdf

This 
```sql
CREATE MATERIALIZED VIEW test.one_materialized_view_v2 on cluster cluster_music_bi_single
AS(
SELECT
    date, hour,app, dime1, dim2,
    sumState(t.valuex) AS value,
    avgState(t.valuex) AS mean_value,
    varPopState(t.valuex) AS var_value
FROM 
(
    SELECT
        date, hour,app, dime1, dim2, userid,
        sum(index_value) AS valuex
    FROM test.one
    GROUP BY date, hour,app, dime1, dim2, userid
) AS t
GROUP BY date, hour,app, dime1, dim2
)
```
does not work by design.

**Comment by ongbo** [comment]
Hi ,@den-crane ，
I understand, is there a solution for demand like me? I want to aggregate to the user's granularity first, and then calculate the average value and variance. If you create a user -grained view view like above, then I will use Tow Group by in your query., But this method will lead to a single -threaded query, which is very slow and it will take 7,8 seconds.This scenario should be common.
Thanks.

**Comment by den-crane** [comment]
There is no magic in this universe.

**Comment by ongbo** [comment]
@den-crane 
IF avg(sum()) is associative.,  sum(sum()^2)) is also not associative. Right?
